critiquebrainz/db/review.py

```diff
@@ -0,0 +1,625 @@
+    """
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT count(*)
+              FROM review
+             WHERE is_draft = :is_drafted
+               AND is_hidden = :is_hidden
+        """), {
+            "is_drafted": is_draft,
+            "is_hidden": is_hidden,
+        })
+        return result.fetchone()[0]
+
+
+def update(*, review_id, drafted, text, license_id=None,
+           language=None, is_draft=None):
+    """Update contents of this review.
+
+    Args:
+        review_id (uuid): ID of the review.
+        drafted (bool): Whether review is presently drafted.
+        license_id(optional, str): Updated license id of a review.
+        is_draft(optional, bool): Whether to publish review or keep it drafted.
+        text(optional, str): Updated text of a review.
+    Returns:
+        Dictionary containing properties of updated review.
+        {
+            "id": uuid,
+            "entity_id": uuid,
+            "entity_type": str("release group", "event", "place"),
+            "user_id": uuid,
+            "user": dict,
+            "edits": int,
+            "is_draft": bool,
+            "is_hidden": bool,
+            "language": str,
+            "license_id": str,
+            "source": str,
+            "source_url": str,
+            "last_revision: dict,
+            "votes": dict,
+            "rating": int,
+            "text": str,
+            "created": datetime,
+            "license": dict,
+        }
+    """
+    updates = []
+    updated_info = {}
+
+    if license_id is not None:
+        if not drafted: # If trying to convert published review into draft.
+            raise BadRequest(lazy_gettext("Changing license of a published review is not allowed."))
+        updates.append("license_id = :license_id")
+        updated_info["license_id"] = license_id
+
+    if language is not None:
+        updates.append("language = :language")
+        updated_info["language"] = language
+
+    if is_draft is not None:
+        if not drafted and is_draft: # If trying to convert published review into draft
+            raise BadRequest(lazy_gettext("Converting published reviews back to drafts is not allowed."))
+        updates.append("is_draft = :is_draft")
+        updated_info["is_draft"] = is_draft
+
+    setstr = ", ".join(updates)
+    query = sqlalchemy.text("""
+        UPDATE review
+           SET {setstr}
+         WHERE id = :review_id
+    """.format(setstr=setstr))
+
+    if setstr:
+        updated_info["review_id"] = review_id
+        with db.engine.connect() as connection:
+            connection.execute(query, updated_info)
+    # Create new revision
+    new_revision = db_revision.create(review_id, text)
+    review = get_by_id(review_id)
+    cache.invalidate_namespace(REVIEW_CACHE_NAMESPACE)
+    return review
+
+
+def create(*, release_group=None, entity_id=None,
+           entity_type=None, user_id, is_draft, text,
```

Comment:
I think this is the perfect time to refactor uses of this function a bit.

1. `entity_type` and `entity_id` arguments shouldn't be optional.
2. `release_group` argument should be removed and its uses replaced by `entity_id` with `entity_type` set to `release_group`.